From 3ed52f8e6f806afbc0894e2f874d33fb90ccbc31 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Tue, 30 Mar 2021 15:52:18 +0200
Subject: [PATCH] PONRTSYS-8628: Remove sleep from RCU reset

Do not sleep, but do a busy polling in the RCU driver when reading back
the RCU status register, when triggered for a system reset. The normal
IP core reset call will still sleep in between.

This is getting called in atomic context when the reboot syscall is
called. When we use the regmap-icc driver to talk to the TEP we have to
sleep in other parts anyway, but when the RCU registers are memory
mapped and directly accessed we do not have to sleep in the other calls
and should also avoid it here. When sleep_us is set to 0, it will not
call usleep_range() at all and the call will not sleep. This should fix
some ugly warning messages seen in the reboot.

This is one of the call traces seen when the kernel checks for correct
handling of atomic contexts.
[ 260.474045] [<600ae590>] ___might_sleep+0x8c/0x3d4
[ 260.478832] [<604b8d44>] intel_assert_device+0x110/0x1f0
[ 260.484106] [<604b8e94>] intel_reset_restart_handler+0x70/0xb4
[ 260.489920] [<600a5480>] notifier_call_chain+0xc8/0x130
[ 260.495129] [<600a5b68>] __atomic_notifier_call_chain+0x140/0x22c
[ 260.501206] [<600a5c68>] atomic_notifier_call_chain+0x14/0x20
[ 260.506936] [<60031c9c>] machine_restart+0x7c/0xb8
[ 260.511710] [<600a879c>] SyS_reboot+0x19c/0x2b4
[ 260.516237] [<6003de38>] syscall_common+0x34/0x58

We could also backport the regmap_read_poll_timeout_atomic() function
from this commit:
https://git.kernel.org/linus/50816a4c39263913d8cfd1ee32f90102679606c6
---
 drivers/reset/reset-intel-syscon.c | 18 +++++++++++++-----
 1 file changed, 13 insertions(+), 5 deletions(-)

diff --git a/drivers/reset/reset-intel-syscon.c b/drivers/reset/reset-intel-syscon.c
index f028d1240796..79ae0fe4ebf0 100644
--- a/drivers/reset/reset-intel-syscon.c
+++ b/drivers/reset/reset-intel-syscon.c
@@ -100,8 +100,8 @@ static u32 intel_stat_reg_off(struct intel_reset_data *data, u32 req_off)
 		return req_off + 0x4;
 }
 
-static int intel_assert_device(struct reset_controller_dev *rcdev,
-			       unsigned long id)
+static int intel_assert_device_timeout(struct reset_controller_dev *rcdev,
+				       unsigned long id, unsigned long sleep_us)
 {
 	int ret;
 	u32 val;
@@ -119,7 +119,13 @@ static int intel_assert_device(struct reset_controller_dev *rcdev,
 
 	regstoff = intel_stat_reg_off(data, regoff);
 	return regmap_read_poll_timeout(data->regmap, regstoff, val,
-					!!(val & BIT(regbit)), 20, 200);
+					!!(val & BIT(regbit)), sleep_us, 200);
+}
+
+static int intel_assert_device(struct reset_controller_dev *rcdev,
+			       unsigned long id)
+{
+	return intel_assert_device_timeout(rcdev, id, 20);
 }
 
 static int intel_deassert_device(struct reset_controller_dev *rcdev,
@@ -208,8 +214,10 @@ static int intel_reset_restart_handler(struct notifier_block *nb,
 
 	for (cnt = 0; cnt < reset_data->rst_num; cnt++) {
 		if (reset_data->rst_ids[cnt]) {
-			intel_assert_device(&reset_data->rcdev,
-					    reset_data->rst_ids[cnt]);
+			/* This is called in atomic context, do not sleep */
+			intel_assert_device_timeout(&reset_data->rcdev,
+						    reset_data->rst_ids[cnt],
+						    0);
 		}
 	}
 
